go/WorkQueue.go

```diff
@@ -120,6 +122,77 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
 	return err
 }
 
+// **************-This method can be used only when database is locked-**************
+func (workQueue *WorkQueue) AddAtomicItem(ctx context.Context, db *redis.Client, item Item) error {
+	txf := func(tx *redis.Tx) error {
+		pipe := db.Pipeline()
+
+		mainQueueLenCmd := pipe.LLen(ctx, workQueue.mainQueueKey)
+		processingQueueLenCmd := pipe.LLen(ctx, workQueue.processingKey)
+
+		_, err := pipe.Exec(ctx)
+
+		mainQueueLen, err := mainQueueLenCmd.Result()
+
+		processingQueueLen, err := processingQueueLenCmd.Result()
+
+		pipeNew := db.Pipeline()
+		processingItemsInQueue, err := pipeNew.LPos(ctx, workQueue.processingKey, item.ID, redis.LPosArgs{
+			Rank:   1,
+			MaxLen: processingQueueLen,
+		}).Result()
+
+		workingItemsInQueue, err := pipeNew.LPos(ctx, workQueue.mainQueueKey, item.ID, redis.LPosArgs{
+			Rank:   1,
+			MaxLen: mainQueueLen,
+		}).Result()
+
+		pipeNew.Exec(ctx)
+
+		if workingItemsInQueue == 0 || processingItemsInQueue == 0 {
+			return nil
+		}
+
+		pipeEx := tx.Pipeline()
+		workQueue.AddItemToPipeline(ctx, pipeEx, item)
+		_, err = pipeEx.Exec(ctx)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	err := db.Watch(ctx, txf, workQueue.mainQueueKey, workQueue.processingKey)
+	//fmt.Println(err)
+	for err == redis.TxFailedErr {
+		if err == nil {
+			// Success.
+			return nil
+		} else if err == redis.TxFailedErr {
+			err = db.Watch(ctx, txf, workQueue.mainQueueKey, workQueue.processingKey)
+		}
+	}
+	return nil
```

Comment:
This could be much more simple:

```go
for {
    err := db.Watch(ctx, txf, workQueue.mainQueueKey, workQueue.processingKey)
    // Retry on transaction failure, return on anything else.
    if err != redis.TxFailedErr {
        return err
    }
}
```

Reply:
<img width="990" alt="image" src="https://github.com/MeVitae/redis-work-queue/assets/69449240/b0b72158-2e79-4ca7-8122-54499f01ee4c">